pkg/apis/deviceplugin/v1/webhook_common.go

```diff
@@ -15,18 +15,139 @@
 package v1
 
 import (
+	"context"
+	"errors"
+	"fmt"
 	"path/filepath"
 	"regexp"
 	"strings"
 
-	"github.com/pkg/errors"
+	"k8s.io/apimachinery/pkg/runtime"
 	"k8s.io/apimachinery/pkg/util/version"
+
+	logf "sigs.k8s.io/controller-runtime/pkg/log"
+	"sigs.k8s.io/controller-runtime/pkg/webhook/admission"
 )
 
 const sha256RE = "@sha256:[0-9a-f]{64}$"
 
+var errObjType = errors.New("invalid object")
+var errValidation = errors.New("invalid resource")
+
 // common functions for webhooks
 
+type commonDevicePluginDefaulter struct {
+	defaultImage string
+}
+
+var _ admission.CustomDefaulter = &commonDevicePluginDefaulter{}
+
+type commonDevicePluginValidator struct {
+	expectedImage     string
+	expectedInitImage string
+	expectedVersion   version.Version
+}
+
+var _ admission.CustomValidator = &commonDevicePluginValidator{}
+
+// Default implements admission.CustomDefaulter so a webhook will be registered for the type.
+func (r *commonDevicePluginDefaulter) Default(ctx context.Context, obj runtime.Object) error {
+	logf.FromContext(ctx).Info("default")
+
+	// type switches can have only one type in a case so the same repeats for
+	// all xDevicePlugin types.
+	// TODO: implement receivers if more complex logic is needed.
+	switch v := obj.(type) {
+	case *DlbDevicePlugin:
+		if len(v.Spec.Image) == 0 {
+			v.Spec.Image = r.defaultImage
+		}
+	case *DsaDevicePlugin:
+		if len(v.Spec.Image) == 0 {
+			v.Spec.Image = r.defaultImage
+		}
```

Comment:
Although there would be more complex logic in future, wouldn't this thing repeated for every plugin still be valid for all of them?   I.e. why not just move it outside the switch?

Reply:
Go's type switch is special and it requires these to be separate types. We also would loose the type of `v` outside the switch. I checked `case *DlbDevicePlugin, *DsaDevicePlugin, *etc...`  which was not allowed either.

Reply:
Thanks for checking!   And I think using generics would complicate this, not simplify it => Resolved.